### PR TITLE
[3251] Add check your information page and create initial allocation request

### LINKED ISSUE #1085: Add Check you information page and create initial allocation request
### Context
As an accredited body, 
I need to review my number before submitting, 
So that I know it is right

```
Given I am on the "How many places would you like to request?" page
When I input a number into the box and click continue
Then I should move to the "Check your information before sending your request" page 

Given I am on the "Check your information before sending your request" page 
When I review the number and click "submit"
Then I should move to the "Request sent confirmation" page 

Given I am on the "Check your information before sending your request" page 
When I click "change" beside the number
Then I should move back to the "How many places would you like to request?" page
```

### Acceptance criteria 
- add "Check your information" page <kbd>
Screenshot 2020-05-07 at 16 38 48</kbd>
- user is able to change the number of places by either clicking Change or Back
- clicking "Send request" 
  - creates an allocation with request_type is "initial" and number of number of places chosen by the user
  - redirects to confirmation page <kbd>
Screenshot 2020-05-07 at 16 39 13</kbd>

### PR DESCRIPTION
**Pull request by AgaDufrat**
### Context
As an accredited body, 
I need to review my number before submitting, 
So that I know it is right

```
Given I am on the "How many places would you like to request?" page
When I input a number into the box and click continue
Then I should move to the "Check your information before sending your request" page 

Given I am on the "Check your information before sending your request" page 
When I review the number and click "submit"
Then I should move to the "Request sent confirmation" page 

Given I am on the "Check your information before sending your request" page 
When I click "change" beside the number
Then I should move back to the "How many places would you like to request?" page
```
Closes #1085
### Changes proposed in this pull request
- add "Check your information" page <kbd>
Screenshot 2020-05-07 at 16 38 48</kbd>
- user is able to change the number of places by either clicking Change or Back
- clicking "Send request" 
  - creates an allocation with request_type is "initial" and number of number of places chosen by the user
  - redirects to confirmation page <kbd>
Screenshot 2020-05-07 at 16 39 13</kbd>

### Guidance to review
- go to http://localhost:3000/organisations/B20/2020/allocations/request
- go through the form
- change number of places
- when you click "Send request"
	- you are taken to the confirmation page
	- allocation exists in the API  (look at `Allocation.last` in the API's console)

### Checklist

- [x] Make sure all information from the Trello card is in here
- [ ] Attach to Trello card
- [ ] Rebased master
- [ ] Cleaned commit history
- [ ] Tested by running locally
- [ ] Product Review

### CODE REVIEW

File: app/views/providers/allocations/number_of_places.html.erb [hunk 1]
```diff
@@ -28,13 +28,18 @@
...
-          <%= label_tag :places, "Say how many trainees the organisation would like to recruit -
+          <%= label_tag :number_of_places, "Say how many trainees the organisation would like to recruit -
             make sure this number is as accurate as possible.",
             class: "govuk-label" %>
-          <%= number_field_tag :places, nil, min: 1, maxlength: 3, class: "govuk-input govuk-input--width-10" %>
+            <%= number_field_tag :number_of_places,
+                    params[:number_of_places],
+                    min: 1,
+                    maxlength: 3,
+                    class: "govuk-input govuk-input--width-10",
```
Comment (aliuk2012) [inline app/views/providers/allocations/number_of_places.html.erb:41]: Could we refactor this to use the form builder?

File: app/views/providers/allocations/check_your_information.html.erb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                    ),
+                    skip_enforcing_utf8: true do |form| %>
+
+      <%= hidden_field_tag "training_provider_code", params[:training_provider_code]  %>
+      <%= hidden_field_tag "number_of_places", params[:number_of_places] %>
+
+      <fieldset class="govuk-fieldset">
+        <%= legend %>
+
+        <%= render "shared/error_wrapper", error_keys: [:check_your_information], data_qa: "allocation__check_your_information" do %>
```
[the remark this answers is not in the document]
Comment (AgaDufrat) [reply app/views/providers/allocations/check_your_information.html.erb:37]: Errors from the API will be displayed here.

[the remark this answers is not in the document]
Comment (aliuk2012) [reply app/views/providers/allocations/check_your_information.html.erb:37]: Ah ok, that makes sense

Actually looking at that partial, all it does is wrap a div around the content and style it like an error if error occured. if we are using the form builder then this shouldn't be needed. Could we use the form builders error summary?

[the remark this answers is not in the document]
Comment (AgaDufrat) [reply app/views/providers/allocations/check_your_information.html.erb:37]: ...but we probably don't need an error wrapper. I won't add it in this PR. Appropriate partial can be added when we are implementing it.

File: app/views/providers/allocations/number_of_places.html.erb [hunk 3]
```diff
@@ -21,25 +21,26 @@
...
-      <%= hidden_field_tag "training_provider_code", params[:training_provider_code]  %>
-
-      <fieldset class="govuk-fieldset">
-        <%= legend %>
-
-        <%= render "shared/error_wrapper", error_keys: [:places], data_qa: "allocation__places" do %>
+                method: :get,
+                builder: GOVUKDesignSystemFormBuilder::FormBuilder do |form| %>
+
+    <%= legend %>
```
Comment (aliuk2012) [inline app/views/providers/allocations/number_of_places.html.erb:27]: We shouldn't need this or the capture block in lines 7-14.

File: app/views/providers/allocations/number_of_places.html.erb [hunk 4]
```diff
@@ -21,25 +21,26 @@
...
+
+    <%= render "shared/error_wrapper", error_keys: [:number_of_places], data_qa: "allocation__places" do %>
```
Comment (aliuk2012) [inline app/views/providers/allocations/number_of_places.html.erb:29]: This shouldn't be needed if we are using the form builder. Just make sure we using the form builders error summary method

File: app/views/providers/allocations/number_of_places.html.erb [hunk 5]
```diff
@@ -21,25 +21,26 @@
...
+      <%= form.govuk_text_field :number_of_places,
+            label: {
+              text: "Say how many trainees the organisation would like to recruit
+                    - make sure this number is as accurate as possible."
```
Comment (aliuk2012) [inline app/views/providers/allocations/number_of_places.html.erb:33]: I don't think it makes sense that the label for the input is the "hint text"

**Comment by AgaDufrat** [comment]
[[M] 8. "Check your information before sending your request" page](https://trello.com/c/WMktYc3R/3251-m-8-check-your-information-before-sending-your-request-page)

File: spec/features/requesting_new_allocations_spec.rb [hunk 6]
```diff
@@ -270,4 +291,45 @@ def when_i_search_for_a_training_provider_with_empty_string
   def and_i_see_error_message_that_i_must_add_more_info
     expect(page).to have_content("You need to add some information")
   end
+
+  def when_i_fill_in_the_number_of_places_input
+    find("#number-of-places-input").fill_in(with: "2")
```
Comment (gpeng) [inline spec/features/requesting_new_allocations_spec.rb:296]: We should use a page object for these so as we're consistent with the other specs. Also happy to discuss whether we still want to use SitePrism but it does stop the HTML structure of the pages leaking across specs which I think is generally a good thing.

File: app/views/providers/allocations/number_of_places.html.erb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    <%= form.govuk_text_field :number_of_places,
+          label: {
+            text: "How many places would you like to request?",
+            size: "xl",
+            tag: "h1",
+          },
+          pattern: "[0-9]*",
+          inputmode: "numeric",
+          width: 10,
+          id: "number-of-places-input" do %>
```
Comment (aliuk2012) [inline app/views/providers/allocations/number_of_places.html.erb:29]: I don't think we need to declare the id. It also breaks the association between the label and the input (i.e try click on the page heading it should set focus on the input)

**Pull request merged**